plugins/field-grid-dropdown/src/index.ts

```diff
@@ -9,169 +9,215 @@
+   *  option & returns a validated language-neutral dropdown option, or null
+   *  to abort the change.
+   * @param config A map of options used to configure the field.
+   *  See the [field creation documentation]{@link
+   * https://developers.google.com/blockly/guides/create-custom-blocks/fields/built-in-fields/dropdown#creation}
+   *  for a list of properties this parameter supports.
+   * @extends {Blockly.Field}
+   * @constructor
    * @throws {TypeError} If `menuGenerator` options are incorrectly structured.
    */
-  constructor(menuGenerator, validator = undefined, config = undefined) {
+  constructor(menuGenerator: Blockly.MenuGenerator,
+      validator?: FieldGridDropdownValidator,
+      config?: FieldGridDropdownConfig) {
     super(menuGenerator, validator, config);
 
-    /**
-     * The number of columns in the dropdown grid. Must be an integer value
-     * greater than 0. Defaults to 3.
-     * @type {number}
-     * @private
-     */
-    this.columns_ = 3;
-    if (config && config['columns']) {
-      this.setColumnsInternal_(config['columns']);
+    if (config?.columns) {
+      this.setColumnsInternal_(config.columns);
     }
 
-    if (config && config['primaryColour']) {
-      this.primaryColour = config['primaryColour'];
+    if (config && config.primaryColour) {
+      this.primaryColour = config.primaryColour;
     }
 
-    if (config && config['borderColour']) {
-      this.borderColour = config['borderColour'];
+    if (config && config.borderColour) {
+      this.borderColour = config.borderColour;
     }
   }
 
   /**
    * Constructs a FieldGridDropdown from a JSON arg object.
-   * @param {!Object} options A JSON object with options.
-   * @returns {!FieldGridDropdown} The new field instance.
+   *
+   * @param config A JSON object with options.
+   * @returns The new field instance.
    * @package
    * @nocollapse
    */
-  static fromJson(options) {
-    // `this` might be a subclass of FieldGridDropdown if that class doesn't
+  static fromJson(config: FieldGridDropdownFromJsonConfig) {
+    if (!config.options) {
+      throw new Error(
+          'options are required for the dropdown field. The ' +
+          'options property must be assigned an array of ' +
+          '[humanReadableValue, languageNeutralValue] tuples.');
+    }
+    // `this` might be a subclass of FieldDropdown if that class doesn't
     // override the static fromJson method.
-    return new this(options['options'], undefined, options);
-  }
-
-  /**
-   * Sets the number of columns on the grid. Updates the styling to reflect.
-   * @param {number} columns The number of columns. Is rounded to
-   *    an integer value and must be greater than 0. Invalid
-   *    values are ignored.
-   * @private
-   */
-  setColumns(columns) {
-    this.setColumnsInternal_(columns);
-    this.updateColumnsStyling_();
+    return new FieldGridDropdown(config.options, undefined, config);
```

Comment:
As per comment, it looks like this should be doing `new this`. Is that possible to maintain?

Reply:
Looks like `this` works fine. Updated!